### PR TITLE
ENH: Knockoff FDR control

### PR DESCRIPTION
**Pull request by kshedden**
This PR implements the knockoff method of Barber and Candes for controlling FDR.  It should have more power and better FDR control than Benjamini & Hochberg when the estimators are dependent, e.g. coefficient estimates in a regression fit.  Another advantage is that it does not require Z-scores or p-values, so is applicable to Lasso and other methods where standard errors are hard to obtain.

http://statweb.stanford.edu/~candes/papers/FDR_regression.pdf

### CODE REVIEW

**Comment by josef-pkt** [comment]
I was looking recently at cases where we don't have reference numbers from other packages and the main verification is from monte carlos.

I think we should include regression tests in those cases, i.e. tests that just use the current numbers as reference numbers, so that we see whether there are problems across platforms and versions and as check for refactorings.
I'm not sure yet what to do with monte carlo tests. Once we start to rely more on those, and there will be more areas where we don't have direct comparisons with other packages, then those could add a lot of time to the test suite. I think at that stage it will be better to keep Monte Carlo tests out of the test suite, but available so we have them for refactorings or if bugs show up. We can rely on regression test for the test runs, as long as we have some initial checks that the function or classes work as intended, e.g. with the Monte Carlo runs.

**Comment by josef-pkt** [comment]
to the topic itself:

I don't quite understand yet how it works based on 15 minute skimming of the article.

`api` question: `knockoff` doesn't sound like a good public module name. Maybe it should be private and have some other public access path. But I have no idea what that should be.

an extension question: If we can add weights/WLS to this, then we could possibly extend it to local approximation in GLM. (If many of the variables don't have a large influence on mu and consequently the weights, then the local approximation might be relatively good.)

**Comment by kshedden** [comment]
Maybe we call it `FDRControl` and try to build an API that would work with Benjamini-Hochberg and other approaches too.

**Comment by kshedden** [comment]
Substantial refactoring here to make this a more open-ended class for doing FDR calculations on regression models.   Right now it only implements the knockoff approach, but the naming and signatures are laid out so that it could incorporate other approaches in the future.  

Test failures seem unrelated to code in this PR.

**Comment by josef-pkt** [comment]
This needs a public API, i.e. which function/class are users supposed to use and what is the import path
e.g. should `RegressionFDR` be imported into the `multitest` module and added to api.py from there, or is one of the knockoff modules public, or is this a regression model?

Given that the emphasis seems to be on FDR, importing through multitest might be the most obvious

File: statsmodels/stats/knockoff_regeffects.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                    x -= np.dot(x, v)*v
+                past.append(x)
+            rv -= np.dot(rv, x) * x
+        z1 = z[0:nvar//2]
+        z2 = z[nvar//2:]
+        st = np.where(z1 > z2, z1, z2) * np.sign(z1 - z2)
+        return st
+
+
+class OLS(RegressionEffects):
```
Comment (josef-pkt) [inline statsmodels/stats/knockoff_regeffects.py]: I would prefer not using OLS given that OLS is already one of our main classes, even if it is imported and used with a path
Is there a postfix that we could add OLSXxx?

**Comment by josef-pkt** [comment]
After skimming and partial reading, but I didn't try to figure out the algorithm or look at the paper again:

This looks now a lot more like statistics based variable selection that we can use for models in either regression or statsmodels.base.
However, I think we can pretty much merge it as is, and "sell" it focused on FDR control and multiple testing. That leaves it open how we tie it in more closely with the models or regression models.

File: statsmodels/stats/knockoff_regeffects.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
+import numpy as np
+import statsmodels.api as sm
+
+
+class RegressionEffects(object):
+    """
+    Prorotype RegressionEffects class.
```
Comment (josef-pkt) [inline statsmodels/stats/knockoff_regeffects.py]: typo Prototype
but maybe better is "Base class for regression effects used in RegressionFDR"

**Comment by josef-pkt** [comment]
looks good

the only question is whether the RegressionEffects need to be in the api, or if they should be imported directly from the module. Adding it to multitests doesn't save on import path length, but it might be useful for keeping things together. This can also be changed later because the effects module is public.

also it needs to be added to the documentation rst file
https://github.com/statsmodels/statsmodels/blame/master/docs/source/stats.rst#L251
(blame seems to be the only display with line numbers)

(file for release notes for 0.9 is still missing, and needs to be added later)

**Pull request merged**